### PR TITLE
Wpt error fixes

### PR DESCRIPTION
**Pull request by zakorgy**
Fixes for the failing WebBluetooth wpt tests, excluding the `disconnect-during` tests cases, due to the lack of the event handling in the current implementation.

Note: The invalid state fixes (1. and 2. commit) should be rethinked, after the event handling is done for disconnections, because it should be handled at the DOM side according to the [Step 6 in the specification](https://webbluetoothcg.github.io/web-bluetooth/#navigating-bluetooth-hierarchy). The current solution is in the `net/bluetooth_thread.rs`, and works only when the adapter is replaced with a different one during testing, because currently calling disconnect does not clear the Maps in the `BluetoothManager`.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

### CODE REVIEW

File: tests/wpt/mozilla/meta/mozilla/bluetooth/readValue/characteristic/characteristic-is-removed.html.ini [hunk 1]
```diff
@@ -1,5 +1,4 @@
 [characteristic-is-removed.html]
   type: testharness
   prefs: [dom.bluetooth.testing.enabled:true]
-  [Characteristic gets removed. Reject with InvalidStateError.]
-    expected: FAIL
+
```
Comment (dati91) [inline tests/wpt/mozilla/meta/mozilla/bluetooth/readValue/characteristic/characteristic-is-removed.html.ini]: remove this line

File: components/net/bluetooth_thread.rs [hunk 2]
```diff
@@ -173,6 +170,13 @@ pub struct BluetoothManager {
     allowed_services: HashMap<String, HashSet<String>>,
 }
 
+enum GATTHierarchyLevel {
+    Device,
+    Service,
+    Characteristic,
+    // Descriptor,
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: dead code in my production code?

File: components/net/bluetooth_thread.rs [hunk 3]
```diff
@@ -242,6 +249,38 @@ impl BluetoothManager {
...
+                    if self.cached_services.contains_key(id) {
+                        if let Some(device_id) = self.service_to_device.get(id) {
+                            return self.is_id_still_in_context(device_id, GATTHierarchyLevel::Device)
+                        }
+                        return false;
+                    }
+                    return false;
+                },
+                GATTHierarchyLevel::Device => {
+                    if self.cached_devices.contains_key(id) {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: I really like this `if something true then return true, otherwise return false` code snippet :)

File: components/net/bluetooth_thread.rs [hunk 4]
```diff
@@ -242,6 +246,38 @@ impl BluetoothManager {
...
+
+    fn is_id_still_in_context(&self, id: &str, id_level: GATTHierarchyLevel) -> bool {
+        loop {
+            match id_level {
+                GATTHierarchyLevel::Characteristic => {
+                    if self.cached_characteristics.contains_key(id) {
+                        if let Some(service_id) = self.characteristic_to_service.get(id) {
+                            return self.is_id_still_in_context(service_id, GATTHierarchyLevel::Service)
+                        }
+                        return false;
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: no need for this if you return false below too

File: components/net/bluetooth_thread.rs [hunk 5]
```diff
@@ -242,6 +246,38 @@ impl BluetoothManager {
...
+                    }
+                    return false;
+                },
+                GATTHierarchyLevel::Service => {
+                    if self.cached_services.contains_key(id) {
+                        if let Some(device_id) = self.service_to_device.get(id) {
+                            return self.is_id_still_in_context(device_id, GATTHierarchyLevel::Device)
+                        }
+                        return false;
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: same in english

File: components/net/bluetooth_thread.rs [hunk 6]
```diff
@@ -242,6 +249,38 @@ impl BluetoothManager {
...
+
+    fn is_id_still_in_context(&self, id: &str, id_level: GATTHierarchyLevel) -> bool {
+        loop {
+            match id_level {
+                GATTHierarchyLevel::Characteristic => {
+                    if self.cached_characteristics.contains_key(id) {
+                        if let Some(service_id) = self.characteristic_to_service.get(id) {
+                            return self.is_id_still_in_context(service_id, GATTHierarchyLevel::Service)
+                        }
+                        return false;
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: we don't need this line

File: components/net/bluetooth_thread.rs [hunk 7]
```diff
@@ -242,6 +249,38 @@ impl BluetoothManager {
         }
     }
 
+
+    fn is_id_still_in_context(&self, id: &str, id_level: GATTHierarchyLevel) -> bool {
+        loop {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: why is this in a loop?

File: components/net/bluetooth_thread.rs [hunk 8]
```diff
@@ -242,6 +249,38 @@ impl BluetoothManager {
         }
     }
 
+
+    fn is_id_still_in_context(&self, id: &str, id_level: GATTHierarchyLevel) -> bool {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: I think you should rethink this method and remove the unnecessary code...

File: components/net/bluetooth_thread.rs [hunk 9]
```diff
@@ -242,6 +246,38 @@ impl BluetoothManager {
...
+                        }
+                        return false;
+                    }
+                    return false;
+                },
+                GATTHierarchyLevel::Device => {
+                    if self.cached_devices.contains_key(id) {
+                        return true;
+                    }
+                    return false;
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: http://static1.fjcdn.com/thumbnails/comments/Top+kek+_c520ddfb9b6f06c9f1efe6bdc01be65c.gif

File: components/net/bluetooth_thread.rs [hunk 10]
```diff
@@ -173,6 +173,13 @@ pub struct BluetoothManager {
     allowed_services: HashMap<String, HashSet<String>>,
 }
 
+enum GATTHierarchyLevel {
+    Device,
+    Service,
+    Characteristic,
+    // Descriptor,
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: What's up with the Descriptors?

File: components/net/bluetooth_thread.rs [hunk 11]
```diff
@@ -790,21 +851,23 @@ impl BluetoothManager {
...
-                                    notify: properties.contains(NOTIFY),
-                                    indicate: properties.contains(INDICATE),
-                                    authenticated_signed_writes: properties.contains(AUTHENTICATED_SIGNED_WRITES),
-                                    reliable_write: properties.contains(RELIABLE_WRITE),
-                                    writable_auxiliaries: properties.contains(WRITABLE_AUXILIARIES),
-                                });
+                if !uuid_is_blacklisted(&uuid, Blacklist::All) {
+                    let properties = self.get_characteristic_properties(&characteristic);
+                    characteristics_vec.push(
+                                    BluetoothCharacteristicMsg {
```
Comment (mmatyas) [inline components/net/bluetooth_thread.rs]: why is this block indented like this?

File: components/script/dom/webidls/BluetoothDevice.webidl [hunk 12]
```diff
@@ -8,9 +8,13 @@
 interface BluetoothDevice {
     readonly attribute DOMString id;
     readonly attribute DOMString? name;
+    // TODO: remove this after BluetoothAdvertisingEvent.
```
Comment (dati91) [inline components/script/dom/webidls/BluetoothDevice.webidl]: maybe add:
... implemented.

File: components/script/dom/bluetooth.rs [hunk 13]
```diff
@@ -4,28 +4,31 @@
 
 use bluetooth_blacklist::{Blacklist, uuid_is_blacklisted};
 use core::clone::Clone;
+use dom::bindings::cell::DOMRefCell;
 use dom::bindings::codegen::Bindings::BluetoothBinding::{self, BluetoothMethods, BluetoothRequestDeviceFilter};
 use dom::bindings::codegen::Bindings::BluetoothBinding::RequestDeviceOptions;
-use dom::bindings::error::Error::{self, Security, Type};
+use dom::bindings::error::Error::{self, Security, Type, NotFound};
```
Comment (mmatyas) [inline components/script/dom/bluetooth.rs]: shouldn't this be `{self, NotFound, Security, Type}`?

File: components/net_traits/bluetooth_scanfilter.rs [hunk 14]
```diff
@@ -69,29 +72,29 @@ impl BluetoothScanfilter {
...
 
     pub fn is_empty_or_invalid(&self) -> bool {
-        (self.name.is_empty() &&
+        (self.name.is_none() &&
          self.name_prefix.is_empty() &&
          self.get_services().is_empty() &&
          self.manufacturer_id.is_none() &&
          self.service_data_uuid.is_empty()) ||
-        self.name.len() > MAX_NAME_LENGTH ||
+        self.get_name().unwrap_or("").len() > MAX_NAME_LENGTH ||
```
Comment (dati91) [inline components/net_traits/bluetooth_scanfilter.rs]: indent

File: tests/wpt/mozilla/meta/mozilla/bluetooth/readValue/characteristic/characteristic-is-removed.html.ini (hunk 1 above)
Comment (mmatyas) [inline tests/wpt/mozilla/meta/mozilla/bluetooth/readValue/characteristic/characteristic-is-removed.html.ini]: purge the heretic

File: tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/filter-matches.html [hunk 15]
```diff
@@ -54,7 +54,6 @@
         test_promises = test_promises
             .then(() => window.navigator.bluetooth.requestDevice(args))
             .then(device => {
-                //assert_in_array(matching_services[0], device.uuids);
```
Comment (mmatyas) [inline tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/filter-matches.html:57]: whats this

File: components/net_traits/bluetooth_scanfilter.rs [hunk 16]
```diff
@@ -69,29 +72,29 @@ impl BluetoothScanfilter {
...
         self.name_prefix.len() > MAX_NAME_LENGTH
     }
 }
 
 #[derive(Deserialize, Serialize)]
-pub struct BluetoothScanfilterSequence(Vec<BluetoothScanfilter>);
+pub struct ScanFilterSequence(Vec<ScanFilter>);
```
Comment (dati91) [inline components/net_traits/bluetooth_scanfilter.rs]: Why did you rename this?

**Review by mmatyas — CHANGES_REQUESTED** [review]
git gud

File: components/net/bluetooth_thread.rs [hunk 17]
```diff
@@ -700,7 +700,10 @@ impl BluetoothManager {
         }
         let mut services_vec = vec!();
         for service in services {
-            if service.is_primary().unwrap_or(false) {
+            if service.is_primary().unwrap_or(false) &&
+            self.allowed_services
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: I think you should move this logic after the for, because this is an unnecessary overhead right now.

File: components/net/bluetooth_thread.rs [hunk 18]
```diff
@@ -777,11 +780,13 @@ impl BluetoothManager {
         let mut services_vec = vec!();
         for service in services {
             if let Ok(service_uuid) = service.get_uuid() {
-                services_vec.push(BluetoothServiceMsg {
-                                      uuid: service_uuid,
-                                      is_primary: service.is_primary().unwrap_or(false),
-                                      instance_id: service.get_id(),
-                                  });
+                if !uuid_is_blacklisted(&service_uuid, Blacklist::All) {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: I think we should move this check out from the for, and check after we have a list.

File: components/net/bluetooth_thread.rs [hunk 19]
```diff
@@ -846,21 +851,23 @@ impl BluetoothManager {
...
-                                    read: properties.contains(READ),
-                                    write_without_response: properties.contains(WRITE_WITHOUT_RESPONSE),
-                                    write: properties.contains(WRITE),
-                                    notify: properties.contains(NOTIFY),
-                                    indicate: properties.contains(INDICATE),
-                                    authenticated_signed_writes: properties.contains(AUTHENTICATED_SIGNED_WRITES),
-                                    reliable_write: properties.contains(RELIABLE_WRITE),
-                                    writable_auxiliaries: properties.contains(WRITABLE_AUXILIARIES),
-                                });
+                if !uuid_is_blacklisted(&uuid, Blacklist::All) {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: I think we should move this check out from the for, and check after we have a list.

File: components/net/bluetooth_thread.rs [hunk 20]
```diff
@@ -911,10 +918,12 @@ impl BluetoothManager {
         let mut descriptors_vec = vec!();
         for descriptor in descriptors {
             if let Ok(uuid) = descriptor.get_uuid() {
-                descriptors_vec.push(BluetoothDescriptorMsg {
-                                         uuid: uuid,
-                                         instance_id: descriptor.get_id(),
-                                     });
+                if !uuid_is_blacklisted(&uuid, Blacklist::All) {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs]: I think we should move this check out from the for, and check after we have a list.

File: components/script/dom/bluetooth.rs [hunk 21]
```diff
@@ -60,6 +69,19 @@ impl Bluetooth {
                            BluetoothBinding::Wrap)
     }
 
+    pub fn get_service_map(&self) -> &DOMRefCell<HashMap<String, MutHeap<JS<BluetoothRemoteGATTService>>>> {
+        &self.service_instance_map
+    }
+
+    pub fn get_characteristic_map(&self)
+                                  -> &DOMRefCell<HashMap<String, MutHeap<JS<BluetoothRemoteGATTCharacteristic>>>> {
```
Comment (dati91) [inline components/script/dom/bluetooth.rs]: this looks odd

File: components/script/dom/bluetooth.rs [hunk 22]
```diff
@@ -44,13 +47,19 @@ const OPTIONS_ERROR: &'static str = "Fields of 'options' conflict with each othe
 pub struct Bluetooth {
     reflector_: Reflector,
     device_instance_map: DOMRefCell<HashMap<String, MutHeap<JS<BluetoothDevice>>>>,
```
Comment (dati91) [inline components/script/dom/bluetooth.rs:50]: this patch should be squashed with [this](https://github.com/szeged/servo/pull/74/commits/a71ae7bc55ad2d5f43996ab5beaaf17a4e66a028)

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 23]
```diff
@@ -123,18 +131,30 @@ impl BluetoothRemoteGATTCharacteristic {
...
             Ok(descriptor_vec) => {
-                Ok(descriptor_vec.into_iter()
-                                 .map(|desc| BluetoothRemoteGATTDescriptor::new(&self.global(),
-                                                                                self,
-                                                                                DOMString::from(desc.uuid),
-                                                                                desc.instance_id))
-                                 .collect())
+                let context = self.service.get().get_device().get_context();
+                for descriptor in descriptor_vec {
+                    if let Some(existing_descriptor) = context.get_descriptor_map()
```
Comment (dati91) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: You can do this without duplicate push.

```
let descriptor = match context... {
    Some => the exsiting desc
    None => create the new desc
}
push desc
```

File: components/script/dom/bluetoothremotegattserver.rs [hunk 24]
```diff
@@ -109,19 +115,30 @@ impl BluetoothRemoteGATTServer {
...
-                                                                             DOMString::from(service.uuid),
-                                                                             service.is_primary,
-                                                                             service.instance_id))
-                              .collect())
+                let context = self.device.get().get_context();
+                for service in service_vec {
+                    if let Some(existing_service) = context.get_service_map()
+                                                           .borrow()
+                                                           .get(&service.instance_id.clone()) {
+                        services.push(existing_service.get());
```
Comment (dati91) [inline components/script/dom/bluetoothremotegattserver.rs]: Same as above

File: components/script/dom/bluetoothremotegattservice.rs [hunk 25]
```diff
@@ -133,7 +146,14 @@ impl BluetoothRemoteGATTService {
         let characteristics_vec = receiver.recv().unwrap();
         match characteristics_vec {
             Ok(characteristic_vec) => {
+                let context = self.device.get().get_context();
                 for characteristic in characteristic_vec {
+                    if let Some(existing_characteristic) = context.get_characteristic_map()
+                                                                  .borrow()
+                                                                  .get(&characteristic.instance_id.clone()) {
+                        characteristics.push(existing_characteristic.get());
```
Comment (dati91) [inline components/script/dom/bluetoothremotegattservice.rs]: Same as above

File: components/net_traits/bluetooth_scanfilter.rs (hunk 16 above)
Comment (zakorgy) [reply components/net_traits/bluetooth_scanfilter.rs]: We use `BluetoothScanfilter` from `net_traits/bluetooth_scanfilter.rs` in `bluetooth.rs` which already has a  `BluetoothScanFilter` in it. So I renamed the `BluetoothScanfilter` and `BluetoothScanfilterSequence` to `ScanFilterSequence` and `ScanFilter`.

**Comment by dati91** [comment]
There should be a list in the PR description (not necessarily here, but for servo sure) about the fixes.

File: components/net_traits/bluetooth_scanfilter.rs (hunk 14 above)
Comment (zakorgy) [reply components/net_traits/bluetooth_scanfilter.rs]: I think this should be in the same row as the `(` in line 75, which get closed in line 79.

File: components/script/dom/bluetooth.rs [hunk 26]
```diff
@@ -6,7 +6,7 @@ use bluetooth_blacklist::{Blacklist, uuid_is_blacklisted};
 use core::clone::Clone;
 use dom::bindings::codegen::Bindings::BluetoothBinding::{self, BluetoothMethods, BluetoothRequestDeviceFilter};
 use dom::bindings::codegen::Bindings::BluetoothBinding::RequestDeviceOptions;
-use dom::bindings::error::Error::{self, Security, Type};
+use dom::bindings::error::Error::{self, Security, Type, NotFound};
```
Comment (zakorgy) [inline components/script/dom/bluetooth.rs]: {self, NotFound, Security, Type}

File: components/net/bluetooth_thread.rs (hunk 11 above)
Comment (fokinv) [reply components/net/bluetooth_thread.rs]: Maybe because the lines are too long otherwise. We should fix this in another patch.

File: tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/filter-matches.html (hunk 15 above)
Comment (zakorgy) [reply tests/wpt/mozilla/tests/mozilla/bluetooth/requestDevice/filter-matches.html:57]: We cant call `device.uuids` here, because the `uuids` idl attribute was not implemented for the `BluetoothDevice` interface, but the chromium test contained it. Meanwhile this `uuids` attribute get removed from the specification so I removed it from the tests.

File: components/script/dom/bluetooth.rs (hunk 21 above)
Comment (fokinv) [reply components/script/dom/bluetooth.rs]: The line would be longer than 120 characters.

File: components/script/dom/bluetoothremotegattcharacteristic.rs [hunk 27]
```diff
@@ -140,16 +140,18 @@ impl BluetoothRemoteGATTCharacteristic {
...
-                                                              .borrow()
-                                                              .get(&descriptor.instance_id.clone()) {
-                        descriptors.push(existing_descriptor.get());
-                        continue;
-                    }
-                    let bt_descriptor = BluetoothRemoteGATTDescriptor::new(&self.global(),
-                                                                           self,
-                                                                           DOMString::from(descriptor.uuid),
-                                                                           descriptor.instance_id.clone());
+                    let bt_descriptor = match context.get_descriptor_map()
```
Comment (dati91) [inline components/script/dom/bluetoothremotegattcharacteristic.rs]: How about you store the list in a var, so you don't have to fetch it every time?

File: components/net/bluetooth_thread.rs [hunk 28]
```diff
@@ -654,7 +615,7 @@ impl BluetoothManager {
                            device_id: String,
                            uuid: String,
                            sender: IpcSender<BluetoothResult<BluetoothServiceMsg>>) {
-        if !self.is_id_still_in_context(&device_id, GATTHierarchyLevel::Device) {
+        if !self.cached_devices.contains_key(&device_id) {
```
Comment (dati91) [inline components/net/bluetooth_thread.rs:618]: Now this turned out to be quite simple :)

File: components/script/dom/bluetooth.rs (hunk 21 above)
Comment (dati91) [reply components/script/dom/bluetooth.rs]: I still suggest you move the `->` to one (or maybe two) tab size from the beginning.

File: components/net_traits/bluetooth_scanfilter.rs (hunk 14 above)
Comment (dati91) [reply components/net_traits/bluetooth_scanfilter.rs]: you're right, but still looks kinda odd :/

**Comment by dati91** [comment]
LGTM
